docs/architecture/adr-068-reverse-sync.md

```diff
@@ -0,0 +1,97 @@
+
+This also means adding new methods to these respective stores in order to persist them
+
+### P2P Light Client Provider
+
+As mentioned previously, the `Dispatcher` is capable of handling requests to multiple peers. We can therefore simply peel off a `blockProvider` instance which is assigned to each peer. By giving it the chain ID, the `blockProvider` is capable of doing a basic validation of the light block before returning it to the client.
+
+It's important to note that because state sync doesn't have access to the evidence channel it is incapable of allowing the light client to report evidence thus `ReportEvidence` is a no op. This is not too much of a concern for reverse sync but will need to be addressed for pure p2p light clients.
+
+### Pruning
+
+A final small note is with pruning. This ADR will introduce changes that will not allow an application to prune blocks that are within the evidence age.
+
+## Future Work
+
+This ADR tries to remain within the scope of extending state sync, however the changes made opens the door for several areas to be followed up:
+- Properly integrate p2p messaging in the light client package. This will require adding the evidence channel so the light client is capable of reporting evidence. We may also need to rethink the providers model (i.e. currently providers are only added on start up)
+- Merge and clean up the tendermint stores (state, block and evidence). This ADR adds new methods to both the state and block store for saving headers, commits and validator sets. This doesn't quite fit with the current struct (i.e. only `BlockMeta`s instead of `Header`s are saved). We should explore consolidating this for the sake of atomicity and the opportunity for batching. There are also other areas for changes such as the way we store block parts. See [here](https://github.com/tendermint/tendermint/issues/5383) and [here](https://github.com/tendermint/tendermint/issues/4630) for more context.
+- Explore opportunistic reverse sync. Technically we don't need to reverse sync if no evidence is observed. I've tried to design the protocol such that it could be possible to move it across to the evidence package if we see fit. Thus only when evidence is seen where we don't have the necessary data, do we perform a reverse sync. The problem with this is that imagine we are in consensus and some evidence pops up requiring us to first fetch and verify the last 10,000 blocks. There's no way a node could do this (sequentially) and vote before the round finishes. Also as we don't punish invalid evidence, a malicious node could easily spam the chain just to get a bunch of "stateless" nodes to perform a bunch of useless work.
```

Comment:
I would say this is a nice-to-have, but might not be worth the engineering effort.

Reply:
Do you mean all of these points or just the last one?

Reply:
Just `opportunistic reverse sync`.

Reply:
yeah I totally agree. Hence why I used the vague term `Explore...` :)

Reply:
Yeah feel free to leave it in then 👍